### PR TITLE
Made it possible to delay the return of a mock or spy.  (#1117)

### PR DESCRIPTION
**Pull request by myrle-krantz**
This is a useful aid in debugging race conditions and other synchronicity problems.

### CODE REVIEW

File: src/main/java/org/mockito/internal/stubbing/answers/AnswersWithDelay.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  private final Answer<Object> answer;
+
+  public AnswersWithDelay(final long sleepyTime, final Answer<Object> answer) {
+    this.sleepyTime = sleepyTime;
+    this.answer = answer;
+  }
+
+  @Override
+  public Object answer(final InvocationOnMock invocation) throws Throwable {
+    TimeUnit.MILLISECONDS.sleep(sleepyTime);
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/stubbing/answers/AnswersWithDelay.java]: Should there also be an async version? Just a thought, not sure if we have to

Comment (myrle-krantz) [inline src/main/java/org/mockito/internal/stubbing/answers/AnswersWithDelay.java]: Which release should I document this as @since ?

Comment (myrle-krantz) [inline src/main/java/org/mockito/internal/stubbing/answers/AnswersWithDelay.java]: Also, by "added in the documentation" where do you mean? In CHANGES.md? or have I missed somewhere else?

Comment (myrle-krantz) [inline src/main/java/org/mockito/internal/stubbing/answers/AnswersWithDelay.java]: I'm not sure what you mean by an "async version"; can you please elaborate?

Without futher details, I'm guessing the answer to the question is "no", since all state in AnswersWithDelay is final.

Though on further thought, it might be interesting to trigger the return externally via another mock method call, rather then make it wait a static time -- that'd be a feature for another time though...

File: src/test/java/org/mockitousage/stubbing/StubbingWithAdditionalAnswersTest.java [hunk 2]
```diff
@@ -86,11 +100,7 @@ public void will_execute_a_void_based_on_strongly_typed_one_parameter_function()
         final IMethods target = mock(IMethods.class);
 
         given(iMethods.simpleMethod(anyString()))
-                .will(answerVoid(new VoidAnswer1<String>() {
-                    public void answer(String s) {
-                        target.simpleMethod(s);
-                    }
-                }));
+                .will(answerVoid((VoidAnswer1<String>) target::simpleMethod));
```
Comment (TimvdLippe) [inline src/test/java/org/mockitousage/stubbing/StubbingWithAdditionalAnswersTest.java]: Please revert these changes. While we build locally on Java 8, we target Java 6 and I would like to make sure it keeps compiling on these older jdks. Our Travis build sadly does not test on jdk 6 any longer as it was deprecated and broken in their containers :sob:

**Comment by myrle-krantz** [comment]
Re: revert. No problem. I was applying the boy scout method and killing off warnings, but I wasn't sure if that change would be acceptable. I could tell you were targeting older java versions, but wasn't sure if that's important in this test too.

I've provided it without the java 8 dependent changes.

**Comment by TimvdLippe** [comment]
Great, this looks good to me. Signing off on this one, but want one other pair of :eyes: to check this PR :smile:

**Comment by matan129** [comment]
IMHO, the PRs title is misleading - in the first place I thought it meant delay on `Mockito.mock()`/`Mockito.spy()`, which is obviously useless. 

Just after reading the code I understood it means answers with delay.
I'd change the title to "Add Delayed Answers".

**Comment by myrle-krantz** [comment]
Can't right now because of day job. Are the release notes drawn from PR titles?

**Comment by szpak** [comment]
I changed a title in #1117. It should be used to generate release notes (as that PR has no labels).

By the way we will test a new CD mechanism - notable version release. Thanks for your contribution @myrle-krantz!

**Comment by szpak** [comment]
My bad, I didn't spot that PR was raised against master not 2.x...

**Comment by szpak** [comment]
Btw, release notes don't look very good. I created #1123.

**Comment by szczepiq** [comment]
Nice contribution. Thank you!

**Comment by myrle-krantz** [comment]
You're welcome, it was my pleasure: Given how intensively mockito helps me every day, this was the least I could do.

Thank you!
Myrle
